File: thirdeye-ui/react-app-old/thirdeye/src/server/routers/actionItems.ts
import { z } from 'zod';
import { router, verifiedUserProcedure, protectedProcedure } from '../trpc';
import { prisma } from '../../lib/prisma';
import { ACTION_ITEM_CATEGORIES, ACTION_ITEM_QUERY_MAPPINGS, SUMMARY_TILE_CONFIG } from '../constants/actionItemCategories';

export const actionItemsRouter = router({
  // Get all action items with real-time database data
  getActionItems: verifiedUserProcedure.query(async () => {
    try {
      // Helper function to convert BigInt to number for JSON serialization
      const convertBigInt = (value: any): number => {
        return typeof value === 'bigint' ? Number(value) : (value || 0);
      };

      // Use action item categories from constants
      const actionItemCategories = ACTION_ITEM_CATEGORIES;

      // Execute all queries in parallel
      const actionItemsWithData = await Promise.all(
        actionItemCategories.map(async (category) => {
          try {
            const result = await prisma.$queryRawUnsafe(category.query) as any[];
            const data = result[0] || { count: 0, cost: 0 };
            
            return {
              id: category.id,
              title: category.title,
              description: category.description,
              subtitle: category.subtitle,
              icon: category.icon,
              color: category.color,
              cost: Math.round(convertBigInt(data.cost)),
              count: convertBigInt(data.count),
              savings: null,
              metadata: {
                type: category.category,
                condition: category.query,
                description: `${category.description} - ${category.subtitle}`,
                action: category.action,
                priority: category.priority
              },
              category: category.category,
              priority: category.priority,
              action: category.action,
              status: 'pending'
            };
          } catch (error) {
            console.error(`Error executing query for ${category.id}:`, error);
            return {
              id: category.id,
              title: category.title,
              description: category.description,
              subtitle: category.subtitle,
              icon: category.icon,
              color: category.color,
              cost: 0,
              count: 0,
              savings: null,
              metadata: {
                type: category.category,
                condition: category.query,
                description: `${category.description} - ${category.subtitle}`,
                action: category.action,
                priority: category.priority
              },
              category: category.category,
              priority: category.priority,
              action: category.action,
              status: 'pending'
            };
          }
        })
      );

      // Calculate total potential savings using the constant query
      const totalPotentialSavings = await (async () => {
        try {
          const result = await prisma.$queryRawUnsafe(SUMMARY_TILE_CONFIG.query) as any[];
          return Math.round(convertBigInt(result[0]?.total_savings));
        } catch (error) {
          console.error('Error calculating total potential savings:', error);
          return 0;
        }
      })();

      // Add summary tile using the constant configuration
      const summaryTile = {
        id: SUMMARY_TILE_CONFIG.id,
        title: SUMMARY_TILE_CONFIG.title,
        description: SUMMARY_TILE_CONFIG.description,
        subtitle: SUMMARY_TILE_CONFIG.subtitle,
        icon: SUMMARY_TILE_CONFIG.icon,
        color: SUMMARY_TILE_CONFIG.color,
        cost: totalPotentialSavings,
        count: 0,
        savings: totalPotentialSavings,
        metadata: {
          type: SUMMARY_TILE_CONFIG.category,
          condition: SUMMARY_TILE_CONFIG.query,
          description: 'Total monthly cost savings from all optimization actions',
          action: SUMMARY_TILE_CONFIG.action,
          priority: SUMMARY_TILE_CONFIG.priority
        },
        category: SUMMARY_TILE_CONFIG.category,
        priority: SUMMARY_TILE_CONFIG.priority,
        action: SUMMARY_TILE_CONFIG.action,
        status: 'pending'
      };

      const allActionItems = [...actionItemsWithData, summaryTile];

      return {
        actionItems: allActionItems,
        totalItems: allActionItems.length,
        pendingItems: allActionItems.filter(item => item.status === 'pending').length,
        totalPotentialSavings: totalPotentialSavings
      };
    } catch (error) {
      console.error('Error fetching action items:', error);
      throw new Error('Failed to fetch action items');
    }
  }),

  // Get action items by category
  getActionItemsByCategory: protectedProcedure
    .input(z.object({
      category: z.enum(['table', 'summary']).optional(),
      priority: z.enum(['high', 'medium', 'low', 'info']).optional(),
      status: z.enum(['pending', 'in_progress', 'completed']).optional()
    }))
    .query(async ({ input }) => {
      try {
        // Get all action items first
        const allActionItemsResult = await actionItemsRouter.createCaller({}).getActionItems();
        let filteredItems = allActionItemsResult.actionItems;

        // Apply filters
        if (input.category) {
          filteredItems = filteredItems.filter(item => item.category === input.category);
        }
        
        if (input.priority) {
          filteredItems = filteredItems.filter(item => item.priority === input.priority);
        }
        
        if (input.status) {
          filteredItems = filteredItems.filter(item => item.status === input.status);
        }

        return {
          actionItems: filteredItems,
          totalItems: filteredItems.length,
          filters: {
            category: input.category,
            priority: input.priority,
            status: input.status
          }
        };
      } catch (error) {
        console.error('Error fetching filtered action items:', error);
        throw new Error('Failed to fetch filtered action items');
      }
    }),

  // Get action item by ID with real-time data
  getActionItemById: protectedProcedure
    .input(z.object({
      id: z.string()
    }))
    .query(async ({ input }) => {
      try {
        // Get all action items and find the requested one
        const allActionItemsResult = await actionItemsRouter.createCaller({}).getActionItems();
        const item = allActionItemsResult.actionItems.find(item => item.id === input.id);
        
        if (!item) {
          throw new Error(`Action item with id '${input.id}' not found`);
        }

        return item;
      } catch (error) {
        console.error('Error fetching action item:', error);
        throw new Error('Failed to fetch action item');
      }
    }),

  // Get detailed table list for a specific action item
  getActionItemTables: protectedProcedure
    .input(z.object({
      actionItemId: z.string(),
      limit: z.number().optional().default(50),
      offset: z.number().optional().default(0)
    }).optional())
    .query(async ({ input }) => {
      try {
        const { actionItemId = 'safe_to_purge', limit = 50, offset = 0 } = input || {};
        
        // Helper function to convert BigInt to number for JSON serialization
        const convertBigInt = (value: any): number => {
          return typeof value === 'bigint' ? Number(value) : (value || 0);
        };

        // Find the action item category configuration
        const actionItemCategory = ACTION_ITEM_CATEGORIES.find(cat => cat.id === actionItemId);
        
        if (!actionItemCategory) {
          throw new Error(`Action item with id '${actionItemId}' not found`);
        }

        // Get the query mapping for detailed table data
        const queryMapping = ACTION_ITEM_QUERY_MAPPINGS[actionItemId];
        
        if (!queryMapping) {
          throw new Error(`No query mapping found for action item '${actionItemId}'`);
        }

        // Build the detailed query to get actual table records
        const detailQuery = `
          SELECT 
            FQN,
            DATABASE_NAME,
            DB_SCHEMA,
            TABLE_NAME,
            COALESCE(SIZE_GB, 0) as SIZE_GB,
            COALESCE(days_since_access, 9999) as days_since_access,
            COALESCE(ROLL_30D_TBL_QC, 0) as ROLL_30D_TBL_QC,
            COALESCE(ROLL_30D_TBL_UC, 0) as ROLL_30D_TBL_UC,
            COALESCE(purge_score, 0) as purge_score,
            COALESCE(monthly_cost_usd, 0) as monthly_cost_usd,
            LAST_ACCESSED_DATE,
            LAST_REFRESHED_DATE
          FROM v_table_purge_scores 
          ${queryMapping.whereClause}
          ${queryMapping.orderClause}
          LIMIT ${limit} OFFSET ${offset}
        `.trim();

        // Get count query for pagination
        const countQuery = `
          SELECT COUNT(*) as total_count 
          FROM v_table_purge_scores 
          ${queryMapping.whereClause}
        `;

        // Debug logging (can be removed in production)
        console.log('🔍 getActionItemTables:', { actionItemId, limit, offset });

        // Execute both queries in parallel
        const [tablesResult, countResult] = await Promise.all([
          prisma.$queryRawUnsafe(detailQuery) as Promise<any[]>,
          prisma.$queryRawUnsafe(countQuery) as Promise<any[]>
        ]);

        // Log results summary
        console.log('📊 Query results:', { tablesCount: tablesResult.length, totalCount: convertBigInt(countResult[0]?.total_count || 0) });

        // Process the results
        const tables = tablesResult.map(table => ({
          FQN: table.FQN || '',
          DATABASE_NAME: table.DATABASE_NAME || '',
          DB_SCHEMA: table.DB_SCHEMA || '',
          TABLE_NAME: table.TABLE_NAME || '',
          SIZE_GB: convertBigInt(table.SIZE_GB),
          days_since_access: convertBigInt(table.days_since_access),
          ROLL_30D_TBL_QC: convertBigInt(table.ROLL_30D_TBL_QC),
          ROLL_30D_TBL_UC: convertBigInt(table.ROLL_30D_TBL_UC),
          purge_score: parseFloat(table.purge_score) || 0,
          monthly_cost_usd: parseFloat(table.monthly_cost_usd) || 0,
          LAST_ACCESSED_DATE: table.LAST_ACCESSED_DATE ? table.LAST_ACCESSED_DATE.toISOString() : null,
          LAST_REFRESHED_DATE: table.LAST_REFRESHED_DATE ? table.LAST_REFRESHED_DATE.toISOString() : null
        }));

        const totalCount = convertBigInt(countResult[0]?.total_count || 0);
        const hasMore = (offset + limit) < totalCount;

        return {
          tables,
          totalCount,
          hasMore,
          actionItem: {
            id: actionItemId,
            title: actionItemCategory.title,
            description: actionItemCategory.description,
            subtitle: actionItemCategory.subtitle,
            category: actionItemCategory.category,
            priority: actionItemCategory.priority,
            action: actionItemCategory.action
          }
        };
      } catch (error) {
        console.error('Error fetching action item tables:', error);
        console.error('Error details:', {
          message: error instanceof Error ? error.message : 'Unknown error',
          stack: error instanceof Error ? error.stack : 'No stack trace',
          actionItemId: input?.actionItemId || 'unknown',
          queryMapping: input?.actionItemId ? ACTION_ITEM_QUERY_MAPPINGS[input.actionItemId] : 'unknown'
        });
        throw new Error(`Failed to fetch action item tables: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
    }),

  // Test endpoint to verify the getActionItemTables logic works
  testActionItemTables: protectedProcedure
    .input(z.object({
      actionItemId: z.string().optional().default('safe_to_purge')
    }).optional())
    .query(async ({ input }) => {
      console.log('testActionItemTables called with input:', input);
      console.log('Input type:', typeof input);
      console.log('Input is undefined:', input === undefined);
      console.log('Input is null:', input === null);
      try {
        // Call the same logic as getActionItemTables but with simpler error handling
        const { actionItemId = 'safe_to_purge' } = input || {};
        console.log('Extracted actionItemId:', actionItemId);
        
        // Helper function to convert BigInt to number for JSON serialization
        const convertBigInt = (value: any): number => {
          return typeof value === 'bigint' ? Number(value) : (value || 0);
        };

        // Find the action item category configuration
        const actionItemCategory = ACTION_ITEM_CATEGORIES.find(cat => cat.id === actionItemId);
        
        if (!actionItemCategory) {
          throw new Error(`Action item with id '${actionItemId}' not found`);
        }

        // Get the query mapping for detailed table data
        const queryMapping = ACTION_ITEM_QUERY_MAPPINGS[actionItemId];
        
        if (!queryMapping) {
          throw new Error(`No query mapping found for action item '${actionItemId}'`);
        }

        console.log('Query mapping found:', queryMapping);

        // Build a simple test query first
        const testQuery = `
          SELECT COUNT(*) as test_count
          FROM v_table_purge_scores 
          ${queryMapping.whereClause}
        `;

        console.log('Executing test query:', testQuery);
        const testResult = await prisma.$queryRawUnsafe(testQuery) as any[];
        console.log('Test query result:', testResult);

        return {
          tables: [],
          totalCount: convertBigInt(testResult[0]?.test_count || 0),
          hasMore: false,
          actionItem: {
            id: actionItemId,
            title: actionItemCategory.title,
            description: actionItemCategory.description,
            subtitle: actionItemCategory.subtitle,
            category: actionItemCategory.category,
            priority: actionItemCategory.priority,
            action: actionItemCategory.action
          },
          debug: {
            queryMapping,
            testResult: testResult[0]
          }
        };
      } catch (error) {
        console.error('Error in test endpoint:', error);
        throw new Error(`Test endpoint failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
    })

});

File: thirdeye-py-service/src/thirdeye/routers/action_items.py
"""
Action Items Router - Cost optimization action items
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text
from typing import Optional
from loguru import logger

from thirdeye.db import get_session
from thirdeye.constants.action_items import (
    ACTION_ITEM_CATEGORIES,
    ACTION_ITEM_QUERY_MAPPINGS,
    SUMMARY_TILE_CONFIG
)

router = APIRouter(prefix="/api/v1/thirdeye/action-items", tags=["action-items"])


def convert_to_number(value) -> float:
    """Convert value to number for JSON serialization"""
    if value is None:
        return 0
    return float(value)


@router.get("")
async def get_action_items(
    session: AsyncSession = Depends(get_session)
):
    """
    Get all action items with real-time database data
    """
    try:
        logger.info("Fetching action items with real-time data")
        
        action_items_with_data = []
        
        # Execute all queries in parallel-like fashion
        for category in ACTION_ITEM_CATEGORIES:
            try:
                result = await session.execute(text(category["query"]))
                row = result.fetchone()
                data = {"count": row.count if row and row.count else 0, "cost": row.cost if row and row.cost else 0}
                
                action_items_with_data.append({
                    "id": category["id"],
                    "title": category["title"],
                    "description": category["description"],
                    "subtitle": category["subtitle"],
                    "icon": category["icon"],
                    "color": category["color"],
                    "cost": round(convert_to_number(data["cost"])),
                    "count": convert_to_number(data["count"]),
                    "savings": None,
                    "metadata": {
                        "type": category["category"],
                        "condition": category["query"],
                        "description": f"{category['description']} - {category['subtitle']}",
                        "action": category["action"],
                        "priority": category["priority"]
                    },
                    "category": category["category"],
                    "priority": category["priority"],
                    "action": category["action"],
                    "status": "pending"
                })
            except Exception as e:
                logger.error(f"Error executing query for {category['id']}: {e}")
                # Add item with zero values on error
                action_items_with_data.append({
                    "id": category["id"],
                    "title": category["title"],
                    "description": category["description"],
                    "subtitle": category["subtitle"],
                    "icon": category["icon"],
                    "color": category["color"],
                    "cost": 0,
                    "count": 0,
                    "savings": None,
                    "metadata": {
                        "type": category["category"],
                        "condition": category["query"],
                        "description": f"{category['description']} - {category['subtitle']}",
                        "action": category["action"],
                        "priority": category["priority"]
                    },
                    "category": category["category"],
                    "priority": category["priority"],
                    "action": category["action"],
                    "status": "pending"
                })
        
        # Calculate total potential savings
        try:
            result = await session.execute(text(SUMMARY_TILE_CONFIG["query"]))
            row = result.fetchone()
            total_potential_savings = round(convert_to_number(row.total_savings if row and row.total_savings else 0))
        except Exception as e:
            logger.error(f"Error calculating total potential savings: {e}")
            total_potential_savings = 0
        
        # Add summary tile
        summary_tile = {
            "id": SUMMARY_TILE_CONFIG["id"],
            "title": SUMMARY_TILE_CONFIG["title"],
            "description": SUMMARY_TILE_CONFIG["description"],
            "subtitle": SUMMARY_TILE_CONFIG["subtitle"],
            "icon": SUMMARY_TILE_CONFIG["icon"],
            "color": SUMMARY_TILE_CONFIG["color"],
            "cost": total_potential_savings,
            "count": 0,
            "savings": total_potential_savings,
            "metadata": {
                "type": SUMMARY_TILE_CONFIG["category"],
                "condition": SUMMARY_TILE_CONFIG["query"],
                "description": "Total monthly cost savings from all optimization actions",
                "action": SUMMARY_TILE_CONFIG["action"],
                "priority": SUMMARY_TILE_CONFIG["priority"]
            },
            "category": SUMMARY_TILE_CONFIG["category"],
            "priority": SUMMARY_TILE_CONFIG["priority"],
            "action": SUMMARY_TILE_CONFIG["action"],
            "status": "pending"
        }
        
        all_action_items = action_items_with_data + [summary_tile]
        
        response = {
            "actionItems": all_action_items,
            "totalItems": len(all_action_items),
            "pendingItems": len([item for item in all_action_items if item["status"] == "pending"]),
            "totalPotentialSavings": total_potential_savings
        }
        
        logger.success(f"Fetched {len(all_action_items)} action items")
        return response
        
    except Exception as e:
        logger.error(f"Error fetching action items: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to fetch action items: {str(e)}")


@router.get("/by-category")
async def get_action_items_by_category(
    category: Optional[str] = Query(None, description="Category filter: table or summary"),
    priority: Optional[str] = Query(None, description="Priority filter: high, medium, low, info"),
    status: Optional[str] = Query(None, description="Status filter: pending, in_progress, completed"),
    session: AsyncSession = Depends(get_session)
):
    """
    Get action items filtered by category, priority, or status
    """
    try:
        logger.info(f"Fetching filtered action items: category={category}, priority={priority}, status={status}")
        
        # Get all action items first
        all_items_response = await get_action_items(session)
        filtered_items = all_items_response["actionItems"]
        
        # Apply filters
        if category:
            filtered_items = [item for item in filtered_items if item["category"] == category]
        
        if priority:
            filtered_items = [item for item in filtered_items if item["priority"] == priority]
        
        if status:
            filtered_items = [item for item in filtered_items if item["status"] == status]
        
        return {
            "actionItems": filtered_items,
            "totalItems": len(filtered_items),
            "filters": {
                "category": category,
                "priority": priority,
                "status": status
            }
        }
        
    except Exception as e:
        logger.error(f"Error fetching filtered action items: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to fetch filtered action items: {str(e)}")


@router.get("/{action_item_id}")
async def get_action_item_by_id(
    action_item_id: str,
    session: AsyncSession = Depends(get_session)
):
    """
    Get a single action item by ID
    """
    try:
        logger.info(f"Fetching action item by ID: {action_item_id}")
        
        # Get all action items and find the requested one
        all_items_response = await get_action_items(session)
        item = next((item for item in all_items_response["actionItems"] if item["id"] == action_item_id), None)
        
        if not item:
            raise HTTPException(status_code=404, detail=f"Action item with id '{action_item_id}' not found")
        
        logger.success(f"Found action item: {action_item_id}")
        return item
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching action item: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to fetch action item: {str(e)}")


@router.get("/{action_item_id}/tables")
async def get_action_item_tables(
    action_item_id: str,
    limit: int = Query(50, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    session: AsyncSession = Depends(get_session)
):
    """
    Get detailed table list for a specific action item with pagination
    """
    try:
        logger.info(f"Fetching tables for action item: {action_item_id}, limit={limit}, offset={offset}")
        
        # Find the action item category
        action_item_category = next((cat for cat in ACTION_ITEM_CATEGORIES if cat["id"] == action_item_id), None)
        
        if not action_item_category:
            raise HTTPException(status_code=404, detail=f"Action item with id '{action_item_id}' not found")
        
        # Get the query mapping
        query_mapping = ACTION_ITEM_QUERY_MAPPINGS.get(action_item_id)
        
        if not query_mapping:
            raise HTTPException(status_code=404, detail=f"No query mapping found for action item '{action_item_id}'")
        
        # Build detail query
        detail_query = text(f"""
            SELECT 
                FQN,
                DATABASE_NAME,
                DB_SCHEMA,
                TABLE_NAME,
                COALESCE(SIZE_GB, 0) as SIZE_GB,
                COALESCE(days_since_access, 9999) as days_since_access,
                COALESCE(ROLL_30D_TBL_QC, 0) as ROLL_30D_TBL_QC,
                COALESCE(ROLL_30D_TBL_UC, 0) as ROLL_30D_TBL_UC,
                COALESCE(purge_score, 0) as purge_score,
                COALESCE(monthly_cost_usd, 0) as monthly_cost_usd,
                LAST_ACCESSED_DATE,
                LAST_REFRESHED_DATE
            FROM v_table_purge_scores 
            {query_mapping['whereClause']}
            {query_mapping['orderClause']}
            LIMIT :limit OFFSET :offset
        """)
        
        # Get count
        count_query = text(f"""
            SELECT COUNT(*) as total_count 
            FROM v_table_purge_scores 
            {query_mapping['whereClause']}
        """)
        
        # Execute both queries
        result = await session.execute(detail_query, {"limit": limit, "offset": offset})
        rows = result.fetchall()
        
        count_result = await session.execute(count_query)
        count_row = count_result.fetchone()
        total_count = int(count_row.total_count if count_row else 0)
        
        # Process tables
        tables = [
            {
                "FQN": row.FQN or "",
                "DATABASE_NAME": row.DATABASE_NAME or "",
                "DB_SCHEMA": row.DB_SCHEMA or "",
                "TABLE_NAME": row.TABLE_NAME or "",
                "SIZE_GB": convert_to_number(row.SIZE_GB),
                "days_since_access": int(row.days_since_access or 9999),
                "ROLL_30D_TBL_QC": int(row.ROLL_30D_TBL_QC or 0),
                "ROLL_30D_TBL_UC": int(row.ROLL_30D_TBL_UC or 0),
                "purge_score": float(row.purge_score or 0),
                "monthly_cost_usd": float(row.monthly_cost_usd or 0),
                "LAST_ACCESSED_DATE": row.LAST_ACCESSED_DATE.isoformat() if row.LAST_ACCESSED_DATE else None,
                "LAST_REFRESHED_DATE": row.LAST_REFRESHED_DATE.isoformat() if row.LAST_REFRESHED_DATE else None
            }
            for row in rows
        ]
        
        has_more = (offset + limit) < total_count
        
        response = {
            "tables": tables,
            "totalCount": total_count,
            "hasMore": has_more,
            "actionItem": {
                "id": action_item_id,
                "title": action_item_category["title"],
                "description": action_item_category["description"],
                "subtitle": action_item_category["subtitle"],
                "category": action_item_category["category"],
                "priority": action_item_category["priority"],
                "action": action_item_category["action"]
            }
        }
        
        logger.success(f"Fetched {len(tables)} tables for action item: {action_item_id}")
        return response
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching action item tables: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to fetch action item tables: {str(e)}")
